[EC-14] Part II: Add Collection Rows to Vault List

jlf0dev opened the pull request:
## Type of change

<!-- (mark with an `X`) -->

```
- [ ] Bug fix
- [x] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

<!--Describe what the purpose of this PR is. For example: what bug you're fixing or what new feature you're adding-->
This PR is for the second half of [EC-14](https://bitwarden.atlassian.net/browse/EC-14), more specifically: 
- [EC-71](https://bitwarden.atlassian.net/browse/EC-71): Add Collections as rows in the vault table
- [EC-62:](https://bitwarden.atlassian.net/browse/EC-62) Update Collection row options
- [EC-60](https://bitwarden.atlassian.net/browse/EC-60): Add Collection column
- [EC-59](https://bitwarden.atlassian.net/browse/EC-59): Create Header

This PR is a follow up to #3440, where we refactored the vault filter extensively to make the collection rows possible. There are some minor UI changes in that PR that I will not repeat here, might be worth a look.

This PR has a server dependency bitwarden/server#2360

Overview of features added:
- Vault table now uses the table in the component library
- Vault table now has a header
    - Header includes a checkbox for select all. In the end user vault, this will select all ciphers. In the org vault, this will select ciphers and collections
    - Header also includes an options button, where you can perform bulk actions on the selected items
- Collections are now rows in the vault _only when a Collection Filter is applied_
- Collections rows are navigatable, you can select a Collection to "drill down" (basically apply that filter)
- Paging in the vault should take into account Collection rows, as well as Cipher rows
- You can perform actions on Collections only in the org vault
    - Deleting is the only option that currently works, Access and Edit Info will be added once [EC-73](https://bitwarden.atlassian.net/browse/EC-73) is finished
- The org vault table now includes a column for "Collections" and "Groups"
    - "Collections" is available when filter by cipher type
    - "Groups" is available when filtering by collection

Known issues:
- ~Access and Edit Info added once [EC-73](https://bitwarden.atlassian.net/browse/EC-73) is finished~
- ~Collections are not currently searchable, they will be hidden once a search is initiated~
- [EC-424](https://bitwarden.atlassian.net/browse/EC-424) will follow up this to remove the leftover Action and New Item menus


## Code changes

<!--Explain the changes you've made to each file or major component. This should help the reviewer understand your changes-->
The main technical ideas include:
- I used the `selectedCollectionNode.children` from the `vault-filter.model` to show the Collection Rows in the vault. This allows me to navigate easily in the vault list by setting the `selectedCollectionNode` to whatever child was selected.
- The organization vault gets the ciphers and collections from the API as opposed to the sync used in the end user vault. I added new endpoint that returns me collections with associated groups (hence the server dependency).
- There is an endpoint added to allow bulk delete of collections (another server dependency).
    - most of the bulk operations logic has been moved into the ciphers component. The bulk actions component will be removed in [EC-424](https://bitwarden.atlassian.net/browse/EC-424)
- I have added badge components for showing the groups and collections in the org vault.
- 

## Screenshots

<!--Required for any UI changes. Delete if not applicable-->
End User Vault:
![image](https://user-images.githubusercontent.com/24985544/197249026-a5f21065-9e32-41b1-8970-e4df0e36fb73.png)

Org Admin Vault:
![image](https://user-images.githubusercontent.com/24985544/197249339-2dd1b814-bc0d-4de3-88da-a608e00092dc.png)

![image](https://user-images.githubusercontent.com/24985544/197249419-161074a6-49ce-46fb-9bbf-626a4c106345.png)


## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

coroiu commented on apps/web/src/app/organizations/vault/ciphers.component.ts:
```diff
@@ -90,16 +147,138 @@ export class CiphersComponent extends BaseCiphersComponent {
 
-  protected showFixOldAttachments(c: CipherView) {
-    return this.organization.canEditAnyCollection && c.hasOldAttachments;
+  selectCollection(c: TreeNode<CollectionFilter>) {
+    this.checkCollection(c);
+  }
```
suggestion: remove function and use `checkCollection` directly

jlf0dev replied:
The organization vault actually has a different behavior when you select a collection versus the end user vault. So I was overriding `selectCollection` here in the org vault and using it to check a collection (a behavior we don't have in the end user vault)

coroiu replied:
Ah right! Ignore this then! :)

coroiu commented on apps/web/src/app/organizations/vault/ciphers.component.ts:
```diff
@@ -90,16 +147,138 @@ export class CiphersComponent extends BaseCiphersComponent {
+  }
+
+  checkCollection(c: TreeNode<CollectionFilter>, select?: boolean) {
+    (c as any).checked = select == null ? !(c as any).checked : select;
```
issue: we should probably avoid casting objects to `any`. Can we solve this in another way? Can we use two separate functions instead? Or maybe check the type with `instanceof` ?

jlf0dev replied:
Introduced `VaultItemRow` to avoid casting to `any`

coroiu commented on apps/web/src/app/organizations/vault/ciphers.component.ts:
```diff
@@ -90,16 +147,138 @@ export class CiphersComponent extends BaseCiphersComponent {
+    this.checkCollection(c);
+  }
+
+  checkCollection(c: TreeNode<CollectionFilter>, select?: boolean) {
+    (c as any).checked = select == null ? !(c as any).checked : select;
+  }
```
suggestion: would it be possible to split this function into two? One called `setCollectionCheck(c, select: boolean)` and one called `toggleCollectionCheck(c)` ?

jlf0dev replied:
I'm not sure more methods would be an ideal situation, but I have simplified it to be `checkRow` instead of having one for each. The logic is also a lot easier to read with null coalescing:
``` javascript
    item.checked = select ?? !item.checked;
```

coroiu commented on apps/web/src/app/organizations/vault/ciphers.component.ts:
```diff
@@ -90,16 +147,138 @@ export class CiphersComponent extends BaseCiphersComponent {
+  }
+
+  getSelectedCollections(): TreeNode<CollectionFilter>[] {
+    return this.activeFilter.selectedCollectionNode?.children.filter((c) => !!(c as any).checked);
```
issue: casting to `any`

jlf0dev replied:
Introduced `VaultItemRow`

coroiu commented on apps/web/src/app/organizations/vault/ciphers.component.ts:
```diff
@@ -90,16 +147,138 @@ export class CiphersComponent extends BaseCiphersComponent {
+    (c as any).checked = select == null ? !(c as any).checked : select;
+  }
+
+  getSelectedCollections(): TreeNode<CollectionFilter>[] {
```
suggestion: convert this to a getter

```
get selectedCollections(): ...
```

jlf0dev replied:
converted

coroiu commented on apps/web/src/app/organizations/vault/ciphers.component.ts:
```diff
@@ -90,16 +147,138 @@ export class CiphersComponent extends BaseCiphersComponent {
+    return this.activeFilter.selectedCollectionNode?.children.filter((c) => !!(c as any).checked);
+  }
+
+  getSelectedCollectionIds(): string[] {
```
suggestion: same here, convert to getter

jlf0dev replied:
converted

coroiu commented on apps/web/src/app/organizations/vault/ciphers.component.ts:
```diff
@@ -90,16 +147,138 @@ export class CiphersComponent extends BaseCiphersComponent {
+    return this.getSelectedCollections()?.map((c) => c.node.id);
+  }
+
+  async editCollectionInfo(c: CollectionView) {
+    return;
+  }
+
+  async editCollectionAccess(c: CollectionView) {
+    return;
+  }
```
question: I'm assuming these are placeholder for my own dialog, but do we need 2 placeholders? Shouldn't it be the same dialog?

jlf0dev replied:
I think the UI has two buttons, one for accessing each tab. I guess it could be a single method with a parameter though.

coroiu commented on apps/web/src/app/organizations/vault/collection-badge/collection-name.badge.component.ts:
```diff
@@ -0,0 +1,24 @@
+  selector: "app-collection-badge",
+  templateUrl: "collection-name-badge.component.html",
+})
+export class CollectionNameBadgeComponent {
```
I think @shane-melton has implemented this exact same component but added it to the CL.
See https://github.com/bitwarden/clients/pull/3563/files#diff-c382ceccac4254a27dc9c5ee5242e0ce45671e7e46a33116d43ab0326ef7a49fR9

suggestion: talk with Shane  😅

coroiu reviewed (commented):
I've only made it through half of the PR but I wanted to submit my comments before continuing. 

One thought I had in general is the handling of collections inside the ciphers component. We're handling them completely separately which results in a few edge cases that we need to add complex code for. Couldn't we add an encapsulation object that supports both collections and ciphers that we can use to display each row instead? That way we won't have to write paging logic or calculate selection logic  for two different types of objects.

I was thinking about something along the line of 

```
interface CipherRow {
  name: string;
  selected: boolean;
  value: CipherView | CollectionView;
}
```

coroiu commented on apps/web/src/app/organizations/vault/ciphers.component.ts:
```diff
@@ -90,16 +147,138 @@ export class CiphersComponent extends BaseCiphersComponent {
-    if (!this.organization.canEditAnyCollection) {
-      return super.deleteCipher(id, this.deleted);
+  protected showFixOldAttachments(c: CipherView) {
+    return this.organization?.canEditAnyCollection && c.hasOldAttachments;
```
That's probably a bad idea which is gonna come back to bite us when we turn on strict mode. Ignore the suggestion.

jlf0dev commented on apps/web/src/app/organizations/vault/ciphers.component.ts:
```diff
@@ -13,19 +22,29 @@ import { StateService } from "@bitwarden/common/abstractions/state.service";
-  organization: Organization;
+  groups: GroupResponse[] = [];
   accessEvents = false;
+  showOrganizationBadge = false;
```
good catch, removed.

jlf0dev commented on apps/web/src/app/organizations/vault/ciphers.component.ts:
```diff
@@ -47,38 +68,74 @@ export class CiphersComponent extends BaseCiphersComponent {
+  }
+
   async load(filter: (cipher: CipherView) => boolean = null, deleted = false) {
     this.deleted = deleted || false;
```
replaced

jlf0dev commented on apps/web/src/app/organizations/vault/vault-filter/vault-filter.service.ts:
```diff
@@ -9,17 +9,29 @@ import { I18nService } from "@bitwarden/common/abstractions/i18n.service";
-  protected collectionViews$ = new ReplaySubject<CollectionView[]>(1);
+export class VaultFilterService extends BaseVaultFilterService implements OnDestroy {
+  private destroy$ = new Subject<void>();
+  private _collections = new ReplaySubject<CollectionGroupDetailsView[]>(1);
```
A behavior subject must have an initial value, this won't be set until we set the org and call the api.

coroiu replied:
https://medium.com/javascript-everyday/behaviorsubject-vs-replaysubject-1-beware-of-edge-cases-b361153d9ccf

I don't think we're negatively impacted by the slightly different behavior (at the moment), but I think it would be nice just out of consistency.

If we're worried about the initial `null` value then we can filter it out

```ts
private _collections = new ReplaySubject<...>(1);
collections$ = this._collections.pipe(filter(c => c != null));
```

If I haven't convinced you with this message then I will drop my case 😁

jlf0dev commented on apps/web/src/app/organizations/vault/vault-filter/vault-filter.service.ts:
```diff
@@ -40,67 +52,53 @@ export class VaultFilterService extends BaseVaultFilterService {
+  protected async loadCollections(org: Organization): Promise<CollectionGroupDetailsView[]> {
+    let collections: CollectionGroupDetailsView[] = [];
+    if (org?.permissions && org?.canViewAssignedCollections) {
+      const collectionResponse = await this.apiService.getManyCollectionsWithDetails(org.id);
```
Agreed, there's a lot that will need to be moved over following our work here.

jlf0dev commented:
Started addressing some of the feedback.

Just so you're aware, I also did some work on removing nested subscriptions in [0bce28f](https://github.com/bitwarden/clients/pull/3875/commits/0bce28fcdca379a1f159a6cd1c1188947d3f0966). It was a pretty self contained change.

jlf0dev commented on apps/web/src/app/organizations/vault/ciphers.component.ts:
```diff
@@ -47,38 +68,74 @@ export class CiphersComponent extends BaseCiphersComponent {
   async applyFilter(filter: (cipher: CipherView) => boolean = null) {
-    if (this.organization.canViewAllCollections) {
+    if (this.organization?.canViewAllCollections) {
       await super.applyFilter(filter);
     } else {
       const f = (c: CipherView) =>
-        c.organizationId === this.organization.id && (filter == null || filter(c));
+        c.organizationId === this.organization?.id && (filter == null || filter(c));
       await super.applyFilter(f);
```
I _wish_ I could prove that the apply filter isn't called before an organization is loaded. To be fair, the only route that will kick this off from outside the component is a `reload()` coming from the `vault.component`, but I can't prove that it won't happen. Something is very wrong if we don't load the organization before everything else though.

I could add a guard that returns early here? What were you imagining it would look like?

coroiu commented:
> I very much agree with your idea, but I am reluctant to make a view model change like that when we're close to a deadline. I'll take a look though and see how involved it would be.

I was thinking the type would be entirely contained inside the cipher component, so no repo-wide view model changes

coroiu commented on apps/web/src/app/organizations/vault/ciphers.component.ts:
```diff
@@ -47,38 +68,74 @@ export class CiphersComponent extends BaseCiphersComponent {
-    this.loaded = true;
+  }
+
+  async refreshCollections(): Promise<void> {
+    await this.vaultFilterService.reloadCollections();
+    if (this.activeFilter.selectedCollectionNode) {
+      this.activeFilter.selectedCollectionNode =
+        await this.vaultFilterService.getCollectionNodeFromTree(
+          this.activeFilter.selectedCollectionNode.node.id
+        );
+    }
```
> I'm hoping this particular method (refreshing the current collection) might be fixed once we transition the model to be an observable, that way collection changes will bubble up and the UI will update.

Yeah I was thinking the same thing, ok let's leave it like it is for now then!

coroiu commented on apps/web/src/app/vault/ciphers.component.html:
```diff
@@ -1,21 +1,163 @@
+        <tr bitRow *ngFor="let col of filteredCollections">
+          <td bitCell class="tw-w-10" (click)="selectRow(col)">
+            <input
+              *ngIf="organization && col.node.name !== 'Unassigned'"
```
question: Will `col.node.name !== 'Unassigned'"` work on vaults with other languages?

jlf0dev replied:
Nice catch, forgot to come back and clean this up 😅

coroiu commented on apps/web/src/app/vault/ciphers.component.html:
```diff
@@ -45,23 +187,31 @@
+          <ng-container *ngIf="!organization">
+            <app-org-badge
+              organizationName="{{ c.organizationId | orgNameFromId: organizations }}"
+              profileName="{{ profileName }}"
```
nit:
```suggestion
              [profileName]="profileName"
```

coroiu commented on apps/web/src/app/vault/ciphers.component.html:
```diff
@@ -45,23 +187,31 @@
+            <app-org-badge
+              organizationName="{{ c.organizationId | orgNameFromId: organizations }}"
+              profileName="{{ profileName }}"
+              (onOrganizationClicked)="onOrganizationClicked(c.organizationId)"
```
thought(non-blocking): this should probably called `(click)` but it's not a change from this PR so don't worry about it 🙂

coroiu commented on apps/web/src/app/organizations/vault/collection-badge/collection-name.badge.component.ts:
```diff
@@ -0,0 +1,24 @@
+  templateUrl: "collection-name-badge.component.html",
+})
+export class CollectionNameBadgeComponent {
+  @Input() collectionIds: string[];
+  @Input() collections: CollectionView[];
```
question: why do we need both `collectionIds` and `collections`? Doesn't the `CollectionView` contain ids?

coroiu commented on apps/web/src/app/vault/pipes/get-collection-name.pipe.ts:
```diff
@@ -0,0 +1,13 @@
+})
+export class GetCollectionNameFromIdPipe implements PipeTransform {
+  transform(value: string, collections: CollectionView[]) {
+    return collections.find((o) => o.id === value)?.name;
```
thought: This is quite inefficient, but for only a few collections the impact is probably negligible. Not sure if there are any organizations with hundreds of collections and/or groups paired with many ciphers that are assigned to many of the previously mentioned collections and/or groups.

coroiu commented on apps/web/src/locales/en/messages.json:
```diff
@@ -563,6 +566,12 @@
     "message": "Folder deleted"
   },
+  "editInfo": {
+    "message": "Edit Info"
```
question: Aren't we moving away from sentence case?
```suggestion
    "message": "Edit info"
```

coroiu commented on libs/common/src/abstractions/organization/organization.service.abstraction.ts:
```diff
@@ -30,7 +30,8 @@ export function canAccessOrgAdmin(org: Organization): boolean {
     canAccessBillingTab(org) ||
-    canAccessSettingsTab(org)
+    canAccessSettingsTab(org) ||
+    org.isManager
```
question: do we need to modify this in any way given the discussions we've had about this?
eg.
https://github.com/bitwarden/clients/pull/3913#pullrequestreview-1160124167

coroiu replied:
Resolving as change was reverted

coroiu reviewed (changes requested):
I've gone through the whole PR now! Ping me when the changes are ready to be reviewed :)

jlf0dev commented on apps/web/src/app/organizations/vault/ciphers.component.ts:
```diff
@@ -90,16 +147,138 @@ export class CiphersComponent extends BaseCiphersComponent {
+      return;
+    }
+
+    const [modal] = await this.modalService.openViewRef(
```
Yep! Refactored each bulk action to be a dialog that uses the dialog service now!

coroiu replied:
And it's looking good!!

jlf0dev commented on apps/web/src/app/organizations/vault/vault-filter/vault-filter.component.ts:
```diff
@@ -29,7 +29,9 @@ export class VaultFilterComponent extends BaseVaultFilterComponent implements On
     this.filters = await this.buildAllFilters();
     if (!this.activeFilter.selectedCipherTypeNode) {
-      this.applyCollectionFilter((await this.getDefaultFilter()) as TreeNode<CollectionFilter>);
+      this.activeFilter.resetFilter();
```
Each vault (org vault/end user vault) has a default filter selected. This was my (less than ideal honestly 😅 ) way of setting the filter when the page loads.

We need a better way to do this but I haven't come up with it yet. Once we integrate the model with the vault filter service I think it will be easier

coroiu replied:
Ok let's leave as-is for now then!

coroiu commented on apps/web/src/app/shared/shared.module.ts:
```diff
@@ -20,6 +20,7 @@ import {
   MenuModule,
   TableModule,
   TabsModule,
+  DialogModule,
```
suggestion(non-blocking): instead of adding the dialog module to the shared module we could instead move each dialog into it's own module and then simply load those modules inside of `VaultModule`. That way we don't increase the size of the shared module and at the same time get rid of stuff from the oversized `loose-components.module` which will help with lazy loading in the future.

So basically we would get a folder for each dialog, and then import them inside the vault-module file.

```
  - web/src/vault
    - bulk-delete-dialog
       bulk-dialog-dialog.module.ts
    - bulk-move-dialog
       bulk-move-dialog.module.ts
     vault.module.ts

jlf0dev replied:
Moved all the dialogs into one module 👍

coroiu reviewed (commented):
New dialogs looking good!

coroiu commented on apps/web/src/app/vault/bulk-delete.component.ts:
```diff
@@ -27,36 +29,66 @@ export class BulkDeleteComponent {
+      deletePromises.push(this.deleteCollections());
     }
 
+    this.formPromise = Promise.all(deletePromises);
```
Which you have now done! Marking this as resolved

coroiu commented on apps/web/src/app/vault/bulk-delete.component.ts:
```diff
@@ -27,36 +29,66 @@ export class BulkDeleteComponent {
 
-  private async deleteCiphersAdmin() {
+  private async deleteCiphersAdmin(): Promise<any> {
     const deleteRequest = new CipherBulkDeleteRequest(this.cipherIds, this.organization.id);
```
In the interest of getting these PRs merge we can view this as a non-blocking suggestion. Feel free to ignore!

jlf0dev commented on apps/web/src/app/vault/bulk-move-dialog.component.ts:
```diff
@@ -0,0 +1,90 @@
+    folderId: ["", [Validators.required]],
+  });
+  folders$: Observable<FolderView[]>;
+  formPromise: Promise<any>;
```
removed

jlf0dev commented on apps/web/src/app/vault/bulk-delete-dialog.component.ts:
```diff
@@ -7,28 +8,66 @@ import { PlatformUtilsService } from "@bitwarden/common/abstractions/platformUti
+  permanent = false;
+  organization: Organization;
 
   formPromise: Promise<any>;
```
Yeah I guess we can remove this, I was planning on taking advantage of it before I started using your handy new `bitAction` 😄

jlf0dev commented on apps/web/src/app/vault/bulk-actions.component.ts:
```diff
@@ -143,26 +125,22 @@ export class BulkActionsComponent {
+      data: { cipherIds: selectedCipherIds },
+    });
+
+    const result = (await lastValueFrom(dialog.closed)) as BulkMoveDialogResult | undefined;
```
Good idea, removed all that

jlf0dev commented on apps/web/src/app/organizations/vault/ciphers.component.ts:
```diff
@@ -151,57 +153,50 @@ export class CiphersComponent extends BaseCiphersComponent implements OnDestroy,
+    const items: VaultItemRow[] = [...this.collections, ...this.ciphers];
+    if (!items) {
       return;
     }
```
javascript 😠

jlf0dev commented on apps/web/src/app/organizations/vault/ciphers.component.ts:
```diff
@@ -130,19 +128,23 @@ export class CiphersComponent extends BaseCiphersComponent implements OnDestroy,
-        c.organizationId === this.organization?.id && (filter == null || filter(c));
-      await super.applyFilter(f);
-    }
+  async load(filter: (cipher: CipherView) => boolean = null, deleted = false) {
+    this.deleted = deleted ?? false;
```
fixed!

jlf0dev replied:
Oh wow so this broke a lot. We're probably passing in a `null` or `undefined` value to `deleted`. We need the default assignment to catch those instances.

coroiu replied:
What how? 😅

jlf0dev commented on apps/web/src/app/vault/ciphers.component.html:
```diff
@@ -1,21 +1,163 @@
+<ng-container *ngIf="filteredCiphers">
+  <bit-table
+    *ngIf="filteredCiphers.length || (filteredCollections?.length && showCollections)"
+    class="!tw-table-fixed"
```
I accidentally left this in there, but its not needed. Originally I wanted to keep the table header visible even if there wasn't any ciphers (and this would stop the columns from changing sizes), but noticed all the other tables would be different. Every other table hides the whole table if its empty. 

I think in the future, maybe VVR, we can change that? I think it would be a better user experience.

jlf0dev commented on apps/web/src/app/vault/ciphers.component.html:
```diff
@@ -1,21 +1,163 @@
-    *ngIf="filteredCiphers.length"
+<ng-container *ngIf="filteredCiphers">
+  <bit-table
+    *ngIf="filteredCiphers.length || (filteredCollections?.length && showCollections)"
```
We actually were able to simplify this to just `filteredCiphers.length || filteredCollections.length` during our pair programming 😄 

(we can also remove the optional chaining from collections since we initialize everything to an empty array now)

jlf0dev commented on apps/web/src/app/organizations/vault/vault.component.ts:
```diff
@@ -73,76 +73,68 @@ export class VaultComponent implements OnInit, OnDestroy {
-              });
-            }
+    if (!this.organization.canUseAdminCollections) {
+      await this.syncService.fullSync(false);
```
Yeah i'm not sure how this is working, but moved below the subscription

jlf0dev commented on apps/web/src/app/organizations/vault/vault.component.ts:
```diff
@@ -73,76 +73,68 @@ export class VaultComponent implements OnInit, OnDestroy {
+          if (
+            // Handle users with implicit collection access since they use the admin endpoint
+            this.organization.canUseAdminCollections ||
+            (await this.cipherService.get(cipherId)) != null
```
moved to switchMap

jlf0dev commented on apps/web/src/app/organizations/vault/vault.component.ts:
```diff
@@ -73,76 +73,68 @@ export class VaultComponent implements OnInit, OnDestroy {
-          this.activeFilter.buildFilter(),
-          this.activeFilter.isDeleted
-        );
+    // eslint-disable-next-line rxjs-angular/prefer-takeuntil
+    this.route.queryParams.pipe(first()).subscribe((qParams) => {
+      this.ciphersComponent.searchText = this.vaultFilterComponent.searchText = qParams.search;
+    });
```
added to both vaults

jlf0dev commented on apps/web/src/app/organizations/vault/vault.component.ts:
```diff
@@ -73,76 +73,68 @@ export class VaultComponent implements OnInit, OnDestroy {
-          const cipher = this.ciphersComponent.ciphers.filter((c) => c.id === qParams.viewEvents);
-          if (cipher.length > 0) {
-            this.viewEvents(cipher[0]);
+    this.route.queryParams
+      // verify that the organization has been set
+      .pipe(withLatestFrom(this.route.parent.params))
+      // eslint-disable-next-line rxjs-angular/prefer-takeuntil, rxjs/no-async-subscribe
+      .subscribe(async ([qParams, params]) => {
```
added

jlf0dev commented on libs/common/src/misc/serviceUtils.spec.ts:
```diff
@@ -0,0 +1,84 @@
+
+import { ServiceUtils } from "./serviceUtils";
+
+describe("serviceUtils", () => {
```
These tests have previously been reviewed in #4092

coroiu commented on apps/web/src/app/vault/ciphers.component.html:
```diff
@@ -72,84 +71,82 @@
+            {{ col.node.name }}
+          </button>
+        </td>
+        <td bitCell class="tw-w-72" (click)="selectRow(col)">
```
question: How does this work with full width layout? How is it *supposed* to work? 

Should we use percentage widths for the columns sizes in the vault ciphers list? @bitwarden/dept-design

danielleflinn replied:
Yes percentage widths makes sense to me. I think that is how we have the component library tables.

jlf0dev replied:
Removed the hardcoded values and replaced with percentages in [1281e3f](https://github.com/bitwarden/clients/pull/3875/commits/1281e3fa0236a30a7b154e99a76566738b156368)

One thing to note is that this will cause the table columns to change width based on whats visible. The only way to fix this is would be to add `table-layout:fixed` to our component. I think thats outside of the scope of this PR but is something we should consider moving forward to VVR.

coroiu commented on apps/web/src/app/vault/vault.component.ts:
```diff
@@ -97,63 +98,71 @@ export class VaultComponent implements OnInit, OnDestroy {
-            });
+    this.route.queryParams
+      .pipe(
+        first(),
```
question: why only `first` ?

jlf0dev replied:
These are operations that should be done the first time we navigate to the vault, but not every time we change the query parameters (to view a certain cipher, etc.)

coroiu reviewed (commented):
In general I think this PR looks good to go, comments are mostly non-blocking and can be ignored. Otherwise some nitpicks and a question for design regarding the full-width layout

jlf0dev commented on apps/web/src/app/vault/bulk-action-dialogs/bulk-move-dialog/bulk-move-dialog.component.ts:
```diff
@@ -0,0 +1,85 @@
+};
+
+@Component({
+  selector: "app-vault-bulk-move",
```
renamed

coroiu reviewed (approved):
Looks good!
